### ISSUE TITLE
Cannot build both tests and examples on MSVC with pseudo-dependencies enabled

### ISSUE THREAD

**Issue by zao** [STEllAR-GROUP/hpx#3215]
I'm building using VS 2017 version 15.6 with a CMake command lifted from the Appveyor build and tried making it build both tests and examples:

> cmake -H. -B../build-2017-debug -G"%MSVC_GENERATOR%" -A%PLATFORM% -DCMAKE_BUILD_TYPE=%CONFIGURATION% -DCMAKE_TOOLCHAIN_FILE=F:/Stellar/vcpkg/scripts/buildsystems/vcpkg.cmake -DHPX_WITH_GIT_COMMIT=%APPVEYOR_REPO_COMMIT% -DHPX_WITH_PSEUDO_DEPENDENCIES=On -DHPX_WITH_EXAMPLES=On -DHPX_WITH_EXAMPLES_HDF5=Off -DHPX_WITH_TESTS=On -DHPX_WITH_RUNTIME=Off -DHPX_WITH_DEPRECATION_WARNINGS=Off

CMake bails out claiming there's duplicate targets:

     CMake Error at cmake/HPX_AddPseudoTarget.cmake:17 (add_custom_target):
      add_custom_target cannot create target "resource_partitioner" because
      another target with the same name already exists.  The existing target is a
      custom target created in source directory "F:/Stellar/hpx/examples".  See
      documentation for policy CMP0002 for more details.
    Call Stack (most recent call first):
      tests/unit/resource/CMakeLists.txt:47 (add_hpx_pseudo_target)

    CMake Error at cmake/HPX_AddPseudoTarget.cmake:17 (add_custom_target):
      add_custom_target cannot create target "throttle" because another target
      with the same name already exists.  The existing target is a custom target
      created in source directory "F:/Stellar/hpx/examples".  See documentation
      for policy CMP0002 for more details.
    Call Stack (most recent call first):
      tests/unit/resource/CMakeLists.txt:47 (add_hpx_pseudo_target)

Disabling pseudo-dependencies mitigates the problem but removes a bunch of useful build targets like `core`.

**Comment by hkaiser** [comment]
For arcane reasons, all parts of our target names (i.e. `a.b.c`) have to be unique across the whole project (on Windows). So in this example `a`, `b`, and `c` can't be used more than once.

**Comment by msimberg** [comment]
Do any of the other allowed special characters have the same problem (basically `+` or `-`)? In the worst case couldn't we translate the dots to something like `examples__resource_partitioner__simple_resource_partitioner`? A bit ugly but better than not working at all.

**Comment by hkaiser** [comment]
@msimberg The reason for the (arcane) behavior we implemented is that we have to drop most of the target name (leaving only its last part) to avoid for the generated path to become too long (target names are directly used as part of the file names used for the generated build system files). So no, concatenating the parts of a target with `_` or any other character wouldn't help, unfortunately.

**Comment by hkaiser** [comment]
On windows cmake generates project files for Visual Studio where we use a different structuring scheme (the `FOLDER` names): 

![image](https://user-images.githubusercontent.com/568063/40207716-9a5be364-59fb-11e8-89d6-54b9b444f0d8.png)

the pseudo target names and dependencies are not needed. Thus leaving those OFF by default without warning is just fine. That allows to selectively build each target. Also, target (link) dependencies are automatically obeyed.

The only caveat is that we have to make sure all parts of pseudo-dependency name are unique across the whole project.

**Comment by msimberg** [comment]
Hmm, okay. So what I was thinking was that *if* pseudo dependencies are not (very) useful on Windows they can be off by default (as it is now) but also forced to off if someone tries to turn them on (so practically remove that option). I admit that this is not really my area so apologies if this is silly. I'll leave it to you Windows people to decide if anything should be done about this (if not I'd like to close this).

**Comment by hkaiser** [comment]
@msimberg forcing this option to off would break AppVeyor as there we rely on building things from the command line which requires for the targets to be available. It would also break any command line based builds other people might use.

Why do you want to change things? The only caveat we have to be aware of is that all name parts used for our targets have to be unique across the HPX build system. Is that a limitation that seems to be too strict to you?

**Comment by msimberg** [comment]
Just updating this since I tried to look into how pseudodependencies broke again on Windows. Two reasons:

1. the `exe` and `component` suffixes were removed meaning that the hierarchical pseudotargets and actual targets get the same name
2. Modules have pseudotargets in the form `tests.{unit,regressions,performance}.libname` which all get mapped to `libname`

This makes it pretty difficult to reasonably try to support Windows. I know @hkaiser etc. try things out regularly but we can't really give any guarantees unless there's a way around this.
